77 ag test cases component new branch 1

ashberg opened the pull request:
Fixes #77 
Fixes #78 
Fixes #79 
Fixes #181

ashberg committed:
Added AGTestCases to ProjectAdmin.

ashberg committed:
In the process of converting the old AGTestCommand code to accomodate the components that have been created since last working on it.

ashberg committed:
Working on the layout of the AGSuiteSettings component.

ashberg committed:
Added modals to create new suites, cases, and commands once those methods are available.

ashberg committed:
Added separate scrolling for the navigation bar and the tab body content.

ashberg committed:
Placed more repeated styling (among the ag_tests components) into the ag_tests.scss file.

ashberg committed:
Adding tests for ag_suite_settings.

ashberg committed:
Adding tests for the ag_command_settings component.

ashberg committed:
Add test cases to test_ag_command_settings for the 'make_min_value_validator's.

ashberg committed:
Added tests for the ag_suite_panel component.

ashberg committed:
Added tests for the ag_case_panel component.

ashberg committed:
Fixed test cases to reflect the new layout of the ag_suites component.

ashberg committed:
Added the 'instructor_files' and 'expected_student_files' fields to the Project instances in the edit_groups components.

ashberg committed:
Added buttons to navigate to the next and previous commands.

ashberg committed:
Added an invoke_focus method to the ValidatedInput component.

ashberg committed:
Ag_suites now implements all AGTestSuite/Case/Command observers.

ashberg committed:
Fixed a logic error in prev_command_is_available.

ashberg committed:
Swapped all dropdowns in the ag_suites directory for select elements.

ashberg committed:
Fixed the tests for ag_command_settings, ag_suite_settings, ag_case_panel and ag_suite_panel.

ashberg committed:
Trying to figure out how to implement going to the next and previous commands.

ashberg committed:
Updated the modal for adding a test case in ag_suite_panel.

ashberg committed:
Added display: flex to the second outermost div in the validated-input component so that the suffix slot element is centered alongside the validated input.

ashberg committed:
Added the item-to-delete class around the suite/case/command name in the deeletion modals.

ashberg committed:
Added tests for ag_case_settings.

ashberg committed:
Filled in tests in test_ag_suites.ts.

ashberg committed:
Removed console.log statements, comments and introduced the reset_active_indices function to the update_ag_test_suite/case/command functions.

ashberg committed:
Removed unnecessary css classes from files inside ag_suites.

ashberg committed:
Adding the padding back to the tab-body in ag_suite/command_settings. Added check to see if suite had been the active level when clicked on, so that it could be viewed instead of closing automatically.

james-perretta commented on src/components/project_admin/ag_suites/ag_case_panel.vue:
```diff
@@ -0,0 +1,380 @@
+                                           && active_command !== null
+                                           && test_case.ag_test_commands[0].pk
+                                              === active_command.pk},
+                 {'parent-of-active-command': command_is_active_level
```
Rename `command_is_active_level`

james-perretta commented on src/components/project_admin/ag_suites/ag_case_panel.vue:
```diff
@@ -0,0 +1,380 @@
+  <div v-if="test_case !== null">
+    <div :class="['test-case',
+                 {'active-case' : is_active_case},
+                 {'active-single-command': test_case.ag_test_commands.length === 1
```
Naming: prefer using `ag_test_case` instead of `test_case` or `ag_case` (similarly for ag_test_suite and ag_test_command)

james-perretta commented on src/components/project_admin/ag_suites/ag_case_panel.vue:
```diff
@@ -0,0 +1,380 @@
+          <context-menu-item ref="delete_case_menu_item"
+                             @context_menu_item_clicked="$refs.delete_case_modal.open()">
+            <template slot="label">
+              <span id="delete-case-label"> Delete case </span>
```
Change "case" to "test case" in UI

james-perretta commented on src/components/project_admin/ag_suites/ag_case_panel.vue:
```diff
@@ -0,0 +1,380 @@
+<template>
+  <div v-if="test_case !== null">
+    <div :class="['test-case',
+                 {'active-case' : is_active_case},
```
You can use one dictionary instead of one per key

james-perretta commented on src/components/project_admin/ag_suites/ag_case_panel.vue:
```diff
@@ -0,0 +1,380 @@
+        <validated-form id="add-command-form"
+                        autocomplete="off"
+                        spellcheck="false"
+                        @submit.native.prevent="add_command"
```
.native.prevent is no longer needed

james-perretta commented on src/components/project_admin/ag_suites/ag_case_panel.vue:
```diff
@@ -0,0 +1,380 @@
+                          hour: 'numeric', minute: 'numeric', second: 'numeric'};
+  case_settings_form_is_valid = false;
+
+  @Watch('test_case', {deep: true})
```
Remove these two watchers

james-perretta commented on src/components/project_admin/ag_suites/ag_case_panel.vue:
```diff
@@ -0,0 +1,380 @@
+    this.test_suite = new_test_suite;
+  }
+
+  toggle_is_open() {
```
De-couple the open/close logic from selecting the active XX

james-perretta commented on src/components/project_admin/ag_suites/ag_case_panel.vue:
```diff
@@ -0,0 +1,380 @@
+  }
+
+  async delete_ag_case() {
+    await this.test_case!.delete();
```
Do we need this `!`?

james-perretta commented on src/components/project_admin/ag_suites/ag_case_panel.vue:
```diff
@@ -0,0 +1,380 @@
+</script>
+
+<style scoped lang="scss">
+@import url('https://fonts.googleapis.com/css?family=Hind|Poppins');
```
Remove

james-perretta commented on src/components/project_admin/ag_suites/ag_case_settings.vue:
```diff
@@ -0,0 +1,124 @@
+  saving = false;
+  settings_form_is_valid = true;
+
+  @Watch('test_case', {deep: true})
```
Remove

james-perretta commented on src/components/project_admin/ag_suites/ag_command_settings.vue:
```diff
@@ -0,0 +1,814 @@
+            <validated-form id="command-settings-form"
+                            autocomplete="off"
+                            spellcheck="false"
+                            @submit.native.prevent="save_ag_test_command_settings"
```
.native.prevent is no longer needed

james-perretta commented on src/components/project_admin/ag_suites/ag_command_settings.vue:
```diff
@@ -0,0 +1,814 @@
+                        </option>
+
+                        <option :value="StdinSource.instructor_file">
+                          Project file content
```
-> Instructor file

james-perretta commented on src/components/project_admin/ag_suites/ag_command_settings.vue:
```diff
@@ -0,0 +1,814 @@
+                          Project file content
+                        </option>
+
+                        <option :value="StdinSource.setup_stdout">
```
Pretty sure nobody is using this or setup_stderr. Let's remove them from the UI

james-perretta commented on src/components/project_admin/ag_suites/ag_command_settings.vue:
```diff
@@ -0,0 +1,814 @@
+                      </div>
+                    </div>
+
+                    <div class="subtract-points-container">
```
We need some indication that these values are negative

james-perretta commented on src/components/project_admin/ag_suites/ag_command_settings.vue:
```diff
@@ -0,0 +1,814 @@
+                </fieldset>
+              </div>
+
+              <div v-if="d_test_command.expected_stdout_source !== ExpectedOutputSource.none ||
```
Break line before `||`

james-perretta commented on src/components/project_admin/ag_suites/ag_command_settings.vue:
```diff
@@ -0,0 +1,814 @@
+  readonly ExpectedOutputSource = ExpectedOutputSource;
+  readonly ExpectedReturnCode = ExpectedReturnCode;
+
+  @Watch('test_command', {deep: true})
```
Remove `{deep: true}`

james-perretta commented on src/components/project_admin/ag_suites/ag_command_settings.vue:
```diff
@@ -0,0 +1,814 @@
+    }
+  }
+
+  @Watch('test_case', {deep: true})
```
Remove `{deep: true}`

james-perretta commented on src/components/project_admin/ag_suites/ag_command_settings.vue:
```diff
@@ -0,0 +1,814 @@
+    }
+  }
+
+  sort_instructor_files() {
```
Remove (API does this)

james-perretta commented on src/components/project_admin/ag_suites/ag_command_settings.vue:
```diff
@@ -0,0 +1,814 @@
+</script>
+
+<style scoped lang="scss">
+@import url('https://fonts.googleapis.com/css?family=Hind|Poppins');
```
Remove

james-perretta commented on src/components/project_admin/ag_suites/ag_suite_panel.vue:
```diff
@@ -0,0 +1,466 @@
+          <validated-form ref="create_case_form"
+                          autocomplete="off"
+                          spellcheck="false"
+                          @submit.native.prevent="create_case"
```
.native.prevent is no longer needed

james-perretta commented on src/components/project_admin/ag_suites/ag_suite_panel.vue:
```diff
@@ -0,0 +1,466 @@
+
+              <div class="new-command-inputs">
+                <fieldset class="case-modal-fieldset">
+                  <legend class="legend"> {{get_ordinal_num(index)}} </legend>
```
Rename `get_ordinal_num` -> `format_ordinal_num`

james-perretta commented on src/components/project_admin/ag_suites/ag_suite_panel.vue:
```diff
@@ -0,0 +1,466 @@
+import { handle_api_errors_async } from '@/utils';
+import { is_not_empty } from '@/validators';
+
+interface CommandInCase {
```
Rename to `NewCommandFields`

james-perretta replied:
Also make this a class that's default-constructable

james-perretta commented on src/components/project_admin/ag_suites/ag_suite_panel.vue:
```diff
@@ -0,0 +1,466 @@
+  @Prop({required: true, type: AGTestSuite})
+  test_suite!: AGTestSuite;
+
+  @Watch('test_suite', {deep: true})
```
Remove

james-perretta commented on src/components/project_admin/ag_suites/ag_suite_panel.vue:
```diff
@@ -0,0 +1,466 @@
+
+  get is_active_suite() {
+    if (this.active_suite !== null && this.active_suite.pk === this.test_suite.pk) {
+      if (!this.is_open) {
```
Getters shouldn't modify state

james-perretta commented on src/components/project_admin/ag_suites/ag_suite_panel.vue:
```diff
@@ -0,0 +1,466 @@
+    }
+  }
+
+  toggle_is_open() {
```
De-couple open/close from active XX selecting

james-perretta commented on src/components/project_admin/ag_suites/ag_suite_panel.vue:
```diff
@@ -0,0 +1,466 @@
+      this.$emit('update_active_suite', this.test_suite);
+    }
+    this.duplicate_command_name_in_case = false;
+    this.new_commands = [{name: "", cmd: ""}];
```
After addressing earlier comment about making it a class, just construct an instance of it

james-perretta commented on src/components/project_admin/ag_suites/ag_suite_panel.vue:
```diff
@@ -0,0 +1,466 @@
+</script>
+
+<style scoped lang="scss">
+@import url('https://fonts.googleapis.com/css?family=Hind|Poppins');
```
Remove

james-perretta commented on src/components/project_admin/ag_suites/ag_suite_settings.vue:
```diff
@@ -0,0 +1,493 @@
+
+                <div class="section-container">
+                  <fieldset class="fieldset">
+                    <legend class="legend"> Grading Related </legend>
```
We should either come up with a more accurate label or just remove this label?

james-perretta commented on src/components/project_admin/ag_suites/ag_suite_settings.vue:
```diff
@@ -0,0 +1,493 @@
+
+                <div class="section-container">
+                  <fieldset class="fieldset">
+                    <legend class="legend"> Project Files </legend>
```
Project Files -> Instructor Files

james-perretta commented on src/components/project_admin/ag_suites/ag_suite_settings.vue:
```diff
@@ -0,0 +1,493 @@
+  @Prop({required: true, type: Project})
+  project!: Project;
+
+  @Watch('test_suite', {deep: true})
```
Remove `{deep: true}`

james-perretta commented on src/components/project_admin/ag_suites/ag_suite_settings.vue:
```diff
@@ -0,0 +1,493 @@
+    );
+  }
+
+  sort_instructor_files() {
```
Remove

james-perretta commented on src/components/project_admin/ag_suites/ag_suite_settings.vue:
```diff
@@ -0,0 +1,493 @@
+    );
+  }
+
+  sort_student_files() {
```
Remove

james-perretta commented on src/components/project_admin/ag_suites/ag_suite_settings.vue:
```diff
@@ -0,0 +1,493 @@
+</script>
+
+<style scoped lang="scss">
+@import url('https://fonts.googleapis.com/css?family=Hind|Poppins');
```
Remove

james-perretta commented on src/components/project_admin/ag_suites/ag_suite_settings.vue:
```diff
@@ -0,0 +1,493 @@
+@import '@/styles/colors.scss';
+@import '@/styles/components/ag_tests.scss';
+@import '@/styles/forms.scss';
+$current-lang-choice: "Poppins";
```
Remove

james-perretta commented on src/components/project_admin/ag_suites/ag_suites.vue:
```diff
@@ -0,0 +1,608 @@
+                           :project="project">
+        </AGCommandSettings>
+      </div>
+    </div> <!--viewing-window-->
```
Remove closing comment

james-perretta commented on src/components/project_admin/ag_suites/ag_suites.vue:
```diff
@@ -0,0 +1,608 @@
+    ValidatedInput
+  }
+})
+
```
Remove blank line

james-perretta commented on src/components/project_admin/ag_suites/ag_suites.vue:
```diff
@@ -0,0 +1,608 @@
+  new_suite_name = "";
+  adding_suite = false;
+
+  index_active_suite = -1;
```
Rename to `index_of_active_suite` or `active_suite_index`

james-perretta commented on src/components/project_admin/ag_suites/ag_suites.vue:
```diff
@@ -0,0 +1,608 @@
+
+      // if there is a command at the same index in the previous case
+      if (this.index_active_case !== 0) {
+        //
```
Remove empty comment

james-perretta commented on src/components/project_admin/ag_suites/ag_suites.vue:
```diff
@@ -0,0 +1,608 @@
+    return false;
+  }
+
+  get next_command_is_available() {
```
command -> test case

james-perretta commented on src/components/project_admin/ag_suites/ag_suites.vue:
```diff
@@ -0,0 +1,608 @@
+    return false;
+  }
+
+  go_to_next_command() {
```
command -> test case

james-perretta commented on src/components/project_admin/ag_suites/ag_suites.vue:
```diff
@@ -0,0 +1,608 @@
+                               parent_suite);
+  }
+
+  go_to_prev_command() {
```
command -> test case

james-perretta commented on src/components/project_admin/ag_suites/ag_suites.vue:
```diff
@@ -0,0 +1,608 @@
+  async add_suite() {
+    try {
+      this.adding_suite = true;
+      await AGTestSuite.create(this.project.pk, { name: this.new_suite_name });
```
Remove spaces inside curly braces

james-perretta commented on src/components/project_admin/ag_suites/ag_suites.vue:
```diff
@@ -0,0 +1,608 @@
+</script>
+
+<style scoped lang="scss">
+@import url('https://fonts.googleapis.com/css?family=Hind|Poppins');
```
Remove

james-perretta commented on src/components/tooltip.vue:
```diff
@@ -147,6 +147,8 @@ export default class Tooltip extends Vue {
     transition: opacity 0.3s;
     word-wrap: break-word;
     z-index: 1;
+    font-weight: 500;
```
If we really need this change, make a separate PR

james-perretta commented on src/components/validated_input.vue:
```diff
@@ -124,6 +124,11 @@ export default class ValidatedInput extends Vue implements Created, Destroyed {
     this.debounced_enable_warnings = debounce(() => this.d_show_warnings = true, 500);
   }
 
+  invoke_focus() {
```
Rename to `focus`

james-perretta commented on src/styles/components/ag_tests.scss:
```diff
@@ -0,0 +1,182 @@
+$parent-of-active-color: white;
+$suite-in-active-container-color: white;
+
+// Dropdown related ************************************************************
```
These aren't defined somewhere else?

james-perretta replied:
Or are you using "dropdown" to mean something else?

james-perretta commented on src/styles/components/ag_tests.scss:
```diff
@@ -0,0 +1,182 @@
+    width: 0;
+}
+
+.input-tooltip {
```
Is this in forms.scss?

james-perretta commented on src/components/project_admin/ag_suites/ag_suites.vue:
```diff
@@ -0,0 +1,608 @@
+    let parent_suite: AGTestSuite = this.active_suite!;
+    let parent_case: AGTestCase = this.active_case!;
+
+    for (let suite_index = 0; suite_index < this.test_suites.length; ++suite_index) {
```
Write a "find parent test case and suite" function to remove this duplicated loop

james-perretta commented on src/components/project_admin/ag_suites/ag_suites.vue:
```diff
@@ -0,0 +1,608 @@
+                                   parent_case,
+                                   parent_suite);
+      }
+      else {
```
Does it work without this else (and removing the earlier call to reset_active_indices)?

james-perretta commented on src/components/project_admin/ag_suites/ag_suites.vue:
```diff
@@ -0,0 +1,608 @@
+    }
+  }
+
+  update_active_case(ag_case: AGTestCase, parent_suite: AGTestSuite) {
```
What if we had these update_active_xx functions only take in one parameter and then search for the parent(s)? Note that an ag test case knows the pk of its parent (similarly with commands)

james-perretta reviewed (changes requested):
- [x] stdin text, expected stdout/stderr text are allowed to be blank
- [x] ag test suite/case/command settings form tests need basic bindings tests with correct values
- [x] Make sure reactive data members use the `d_` prefix
- [x] fields that start blank but are required should use show_warnings_on_blur=true
- [x] Use a white button for the delete modal cancel buttons
- [x] Use the format date utility function instead of copy-pasting last_modified_format
- [x] In new test case dialog, don't show the "first" command divider unless there's more than one command
- [x] Add tooltip to "Command": "Can be any valid bash command. Note that if it includes sequencing or piping, you will have to increase the process limit."
- [x] Delete test case should be red in test case panel menu
- [x] .native.prevent is no longer needed for validated form
- [x] The create ag test case modal adds a command every time it opens
- [x] Add some icons to the test case panel menu? (plus, pencil, copy, trash)
- [x] In my browser, the next/prev footer cuts off the save button (still slightly cut off, last modified is totally cut off)
- [x] Change "Save updates" to "Save"
- [x] Needs loading icons for v-if content that depends on async requests
- [x] The API sorts instructor files, so you don't need to
- [x] When clicking on an open, active suite or case, it should close the panel, but not change the active_xx (that closed panel should be highlighted). 
- [x] The next and previous buttons should say "next test" and "prev test" instead of "next command"
- [x] Change "Project file content" to "Instructor file"
- [x] Wrongness deductions need some indication that the values are negative
- [x] If a test case is active and you select the suite that test case is in, the suite should become active and stay open
Alternatively, re-order the words so that it's "active level is command" instead of "command is active level"
- [x] In command settings form, hide command name input if it's a single-command test case
- In places where the word "case" is on its own (case settings, case name, etc.) should be replaced with either "test" or "test case"
- helper function to create generic commands could be used in more places (moving it to tests/utils.ts is ok) -- there's an umbrella issue for this
- AG case settings modal should probably close after saving (revisit this after the feedback options are added)
- ~~Rename "active level" to "active type". In places where it's just "level," make it "active type."~~

ashberg committed:
One single function in ag_suites.vue now updates the active suite/command. Added test cases/changed test cases to reflect the changes requested on the pull request.

ashberg committed:
Removed comments and unnecessary html code.

ashberg committed:
Changed 'Project file content' option label in the ag_command_settings component to 'Instructor file content'.

ashberg committed:
Removed setting the active_ag_test_command to null in the update_ag_test_command_deleted function in ag_suites.

ashberg committed:
Created functions to populate the data needed for the ag_suites tests.

ashberg committed:
Renamed the 'update_active_thing' function in ag_suites to 'update_active_item'. Added from_string_fn to every validated input in ag_test_commands where the input should be a number and not a string.

ashberg committed:
Grouped the variables together in ag_suite_panel.

james-perretta committed:
Renamed parameter thing -> item in update_active_item.

james-perretta committed:
Misc changes:
- Minor indentation/wrapping change.
- Simplified/generalized dupllicate_command_name property in ag_suite_panel.vue.
- Removed .native.prevent from ag_suite_settings.vue.
- Removed some spaces inside curly braces.
- Removed client-side sorting of instructor and student files. (the server handles this now)
- Added check to create ag test case tests the makes sure the duplicate command name messages are displayed.

james-perretta committed:
Refactored ag command settings tests to reduce duplication.

james-perretta committed:
Added loading icon to ag_suites.vue.

james-perretta committed:
Fixed some type errors in ag command settings tests and replaced remaining uses of AxiosError mocking with HttpError.

james-perretta committed:
Changed 'Save updates' -> 'Save' on save buttons.

james-perretta committed:
Added column scrolling to ag_suites.vue.
Also adjusted prev/next buttons footer so it doesn't block the bottom of the page.
Added vertical divider between sidebar and main content in ag_suites.vue.

james-perretta committed:
Fixed issue where a new command was added every time the create test case modal opened.
Added tooltip to command input.

james-perretta committed:
Added icons to ag_case_panel.vue context menu. Made the delete test case option red in context menu.

james-perretta commented:
Also fixes #230

james-perretta committed:
Replaced old datetime formatting with format_datetime functions.
Changed color of modal cancel buttons and tweaked the wording for deletion modals.

james-perretta committed:
Added show_warnings_on_blur to some initially-blank inputs.

james-perretta committed:
Renamed reactive variables in ag_suites components.

ashberg committed:
Removed the is_not_empty validator from the stdin/stdout/stderr text inputs in ag_command_settings. Added v-binding tests to inputs in the ag_suites directory. Changed some inputs that were using select tags but v-modeling an object to use the Dropdown component in order to test selecting values in those dropdowns. Implemented the test to make sure a new command doesnt get added when a user closes the modal to create a test case without pressing the Create Case button.

ashberg committed:
Added v-model binding tests for data in the ag_suites directory.
